alphagov/paas-admin
# [#154558689] Implement oauth2

## What

We're implementing a OAuth2 solution in pazmin. We're using `passport` library to handle most of the magic for us. 

We've covered all code used, but there are things we're not testing due to the approach taken in the beginning of this story.

We'd like to test whether a cookie that has been tampered with, does not match with the signature meaning the app would logout the user.

Rafal really doesn't like "magical" middleware that simply applies logic to any route that follows it's declaration. Bother him about details. - This is still on the table and may end up being improved in a follow up story. UNLESS the reviewer agrees with Rafal.

We've also encountered a problem that a thrown error will not stop the tests but instead will hang them. We agreed this will be fixed in a follow up story.

Along there are some typo corrections and cleaning up/fixing some other bugs introduced in previous stories.

## How to review

- Code review
- Run the tests `npm test` (will be run by travis)
- Run the application 
  ```sh
  OAUTH_AUTHORIZATION_URL="https://uaa.rafalp.dev.cloudpipeline.digital/oauth/authorize" OAUTH_TOKEN_URL="https://uaa.rafalp.dev.cloudpipeline.digital/oauth/token" OAUTH_CLIENT_ID="rafal-login" OAUTH_CLIENT_SECRET="secret" SERVER_ROOT_URL="http://localhost:3000" npm start
  ```
- Attempt to preview `/orgs` - you should be redirected to Rafal's UAA
- Ask nicely for the password or an account
- Log in - you should be redirected back to the `/orgs` and see the actual page
- Check if the token is valid and we can pull orgs from CF API - we've prepared a TMP commit 
1f2382e 
- Confirm app displays some orgs user has access to `/tmp-orgs`
- Remove the TMP commit once not needed.

🐝 🐝 🐝 🐝  You should remove the TMP commit. Also, the travis fails purely because the code is not covered in tests.

## Who can review

Neither @keymon, @46bit, @carolinegreen nor myself.
